cw/cw305/capture.py

```diff
@@ -330,6 +330,124 @@ def capture_aes_fvsr_key(ot):
         yield ret
 
 
+def capture_aes_fvsr_key_batch(ot, ktp, capture_cfg, scope_type):
+    """A generator for capturing AES traces for fixed vs random key test in batch mode.
+    The data collection method is based on the derived test requirements (DTR) for TVLA:
+    https://www.rambus.com/wp-content/uploads/2015/08/TVLA-DTR-with-AES.pdf
+    The measurements are taken by using either fixed or randomly selected key in batches.
+    In order to simplify the analysis, the first encryption has to use fixed key.
+    To generate random keys and texts internal PRNG's are used instead of AES functions as specified
+    in the DTR.
+
+    Args:
+      ot: Initialized OpenTitan target.
+      ktp: Key and plaintext generator.
+      capture_cfg: Capture configuration.
+      scope_type: cw or waverunner as a scope for batch capture.
+    """
+    # Seed host's PRNG.
+    # TODO: Replace this with a dedicated PRNG to avoid other packages breaking our code.
+    random.seed(capture_cfg["batch_prng_seed"])
+    # Seed the target's PRNG
+    ot.target.simpleserial_write("s", capture_cfg["batch_prng_seed"].to_bytes(4, "little"))
+
+    key_fixed = bytearray([0x81, 0x1E, 0x37, 0x31, 0xB0, 0x12, 0x0A, 0x78,
+                           0x42, 0x78, 0x1E, 0x22, 0xB2, 0x5C, 0xDD, 0xF9])
+    tqdm.write(f'Fixed key: {binascii.b2a_hex(bytes(key_fixed))}')
+    ot.target.simpleserial_write("t", key_fixed)
+
+    sample_fixed = 1
+
+    # Create the ChipWhisperer project.
+    project = cw.create_project(capture_cfg["project_name"], overwrite=True)
+    # Capture traces.
+    rem_num_traces = capture_cfg["num_traces"]
+    num_segments_storage = 1
+    # cw and waverunner scopes are supported for batch capture.
+    scope = SCOPE_FACTORY[scope_type](ot, capture_cfg)
+
+    with tqdm(total=rem_num_traces, desc="Capturing", ncols=80, unit=" traces") as pbar:
+        while rem_num_traces > 0:
+            # Determine the number of traces for this batch and arm the oscilloscope.
+            scope.num_segments = min(rem_num_traces, scope.num_segments_max)
+            scope.arm()
+
+            # Start batch encryption.
+            ot.target.simpleserial_write("f", scope.num_segments_actual.to_bytes(4, "little"))
+
+            # Transfer traces.
+            waves = scope.capture_and_transfer_waves()
+            assert waves.shape[0] == scope.num_segments
+            # Check that the ADC didn't saturate when recording this batch.
+            check_range(waves, ot.scope.adc.bits_per_sample)
```

Comment:
Does it have any impact on the capture rate if the reading out/transferring of traces is moved down to after computing the ciphertexts? Because while the encryption is running, the host could theoretically do something else instead of waiting for the capture to finish in `scope.capture_and_transfer_waves()`.

Reply:
Yes it also increased the capture rate. I was going to add that in a seperate PR.

Reply:
Sounds good to me.